fix(files): Correctly copy the cache information during copy operations

artonge opened the pull request:
Needed to copy the `encrypted` flag of encrypted files when those files are two level down in a moved folder.

Reproduction steps of the issue:

- Set up the following folder structure
  - `/A/B/t.txt`
  - `/C`
- Copy `A` in `C`
 - `t.txt` is encrypted but does not have the `encrypted` flag set to `true`, and therefore not readable.

With this PR, `t.txt` is readable.

Note: Open to better ways to share code between the two methods.

Added a test to ensure file content stays the same after a copy.

joshtrichards commented:
This might help with #48174.

come-nc reviewed (approved):
Kind of :elephant: , I do not understand what’s going on here

artonge commented:
Snap, the `ObjectStoreStorage` does the cache copy inside the storage copy, so the final cache copy creates another filecache entry.

https://github.com/nextcloud/server/blob/6a0e11783f6fc921a1e6569f72b6dd864d6189cc/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L699-L700

artonge committed:
fix(files): Correctly copy the cache information on copy operations

Needed to copy the `encrypted` flag of encrypted files when those files are two level down in a moved folder.

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
test(View): Copy preserves content

Signed-off-by: Louis Chemineau <louis@chmn.me>

icewind1991 commented:
This introduced a regression for some cross-storage moves, see https://github.com/nextcloud/server/pull/49695

kyteinsky commented:
hello, was this backported to 29/30? Would you mind helping me with the links?

joshtrichards commented:
> hello, was this backported to 29/30? Would you mind helping me with the links?

No, it was not.

nfebe commented:
/backport to stable30

nfebe commented:
/backport to stable29
